### ISSUE TITLE
rerun warning with bbl

### ISSUE THREAD

**Issue by dkwo** [plk/biblatex#1112]
On my system, the sequence pdflatex->biber->pdflatex(x2) produces no warning.
However, if I delete all files and keep just .tex and .bbl files and run pdflatex (several times), I get the warning
```
LaTeX Warning: There were undefined references.

Package biblatex Warning: Please (re)run Biber on the file:
(biblatex)                main
(biblatex)                and rerun LaTeX afterwards.
```
I cannot find differences in the two outputs, although
```
diff main.pdf main2.pdf
Binary files main.pdf and main2.pdf differ
```
The minimal example I'm using
```
\documentclass{article}
\usepackage{biblatex}
\addbibresource{draft.bib}
\begin{document}
\cite{MR710056}
\printbibliography
\end{document}
```
with the bibfile
```
@article {MR710056,
    AUTHOR = {Donaldson, S. K.},
     TITLE = {An application of gauge theory to four-dimensional topology},
   JOURNAL = {J. Differential Geom.},
  FJOURNAL = {Journal of Differential Geometry},
    VOLUME = {18},
      YEAR = {1983},
    NUMBER = {2},
     PAGES = {279--315},
      ISSN = {0022-040X},
   MRCLASS = {57N13 (53C05 57R10 57R55 58E15)},
  MRNUMBER = {710056},
MRREVIEWER = {R. C. Kirby},
       URL = {http://projecteuclid.org/euclid.jdg/1214437665},
}
```
My texlive uses biblatex v3.15a, so perhaps issue #1072 could solve my problem.
Otherwise, do you know what the problem could be?
Thanks.

**Comment by moewew** [comment]
This is not too surprising. Biber rerun detection relies on the `.aux` file. If you clear it, `biblatex` will think you need a rerun. I would not recommend deleting only some of the temporary files: Either keep them all or delete them all. Keeping only some around might lead to a sort of intermediate state where code relies rely on different temporary data.

I can no longer reproduce the behaviour with `biblatex` v3.16, so I'd speculate that the changes for #1072 make your rerun warning go away if you run LaTeX twice.

In general I would not rely on or worry too much about those warnings.

Usually there is a way to make them go away if you run LaTeX and Biber often enough. Sometimes they get things 'wrong' in that they want you to run LaTeX or Biber more often than strictly necessary to obtain converging output. It is never a bad idea to delete all temporary files and then rerun LaTeX, Biber, LaTeX, LaTeX + additional requested runs before you produce the final version of your document, just to be sure that everything is OK.

On the other hand you can also not rely on not needing a Biber run when `biblatex` does not ask for it. `biblatex` does not monitor the contents of your `.bib` file, so if you for example fix a typo in the title of an entry, you still need to manually rerun Biber. (Unless of course you use a fancy tool like `latexmk` that monitors your files.)

**Comment by dkwo** [comment]
Thanks.
In my case I did not edit the .tex nor the .bbl files, just delete everything else.
The use-case I wanted to reproduce is upload to arXiv, where one uploads bbl and tex.
It seems v3.16 removes the warning, so perhaps nothing to worry about.

**Comment by moewew** [comment]
OK, since the warning goes away after running pdfLaTeX the second time this problem appears to be fixed. I didn't conduct a full investigation, but if the issue was present in v3.15a, the most likely explanation is that it was resolved when things were refactored for #1072. So I'm closing this and mark it as a duplicate.
